### ISSUE TITLE
Get plugins by language matching

### ISSUE THREAD

**Issue by TiceCosmos** [dprint/dprint#365]
I use this config file: .clangd, .clang-format, .clang-tidy. 
These files are actually YAML.
I want to use dprint-plugin-prettier to format this file.
Is there any way to deal with it? Not to modify dprint-plugin-prettier.

**Comment by TiceCosmos** [comment]
1. get language by filename and get plugin by language, imaginary configurations:
   ```json
   {
     "includes": ["**/*.{json,jsonc}", "**/.{clangd,clang-format,clang-tidy}"],
     "associations": {
       "*.json": "json",
       "*.jsonc": "json",
       ".clang-format": "yaml",
       ".clangd": "yaml"
     },
     "plugins": {
       "json": "https://plugins.dprint.dev/json-0.11.0.wasm",
       "yaml": "https://plugins.dprint.dev/prettier-0.2.2.exe-plugin@63b06beba3acd51e6d23379c47db342a1a24f3d88f3c52a775e2db3426124582"
     }
   }
   ```

2. combining plugin and includes, imaginary configurations:
   ```json
   {
     "plugins": [
       {
         "includes": ["**/*.{json,jsonc}"],
         "plugin": "https://plugins.dprint.dev/json-0.11.0.wasm"
       },
       {
         "includes": ["**/.{clangd,clang-format,clang-tidy}"],
         "plugin": "https://plugins.dprint.dev/prettier-0.2.2.exe-plugin@63b06beba3acd51e6d23379c47db342a1a24f3d88f3c52a775e2db3426124582"
       }
     ]
   }
   ```

**Comment by dsherret** [comment]
Yes, this would be very good, but I'm not sure what the best way of doing it would be. I need to think about it more.

**Comment by TiceCosmos** [comment]
Compatible with existed configurations:

```jsonc
{
  "$schema": "https://dprint.dev/schemas/v0.json",
  "includes": [
    "**/*.{json,jsonc,yml}"
  ],
  "plugins": [
    "https://plugins.dprint.dev/json-0.11.0.wasm",
    "https://plugins.dprint.dev/prettier-0.2.2.exe-plugin@63b06beba3acd51e6d23379c47db342a1a24f3d88f3c52a775e2db3426124582"
  ],
  // Add the following
  "include_maps": {
    "prettier": // plugin name
    [ // plugin include
      "**/.{clangd,clang-format,clang-tidy}"
    ]
  }
}
```

@dsherret What do you think of this scheme.

**Comment by dsherret** [comment]
Another idea is to introduce a special property on each plugin config. I think I like this one best, but I'm not sure what to name the property:

```jsonc
{
  "json": {
    "associations": "*.rc" // could be a string or an array of globs
  },
  "prettier": {
    "associations": "**/my-prettier-only-folder/*",
  },
  "includes": [
    "**/*.{json,rc,js,ts}"
  ],
  "plugins": {
    "https://plugins.dprint.dev/json-0.11.0.wasm",
    "https://plugins.dprint.dev/prettier-0.2.2.exe-plugin@etc..."
  }
}
```

**Comment by joscha** [comment]
@dsherret can I pull your attention back to this one please? We are trying to write a plugin as outlined in #411 but matching the plugin to the files it has to format is a real blocker for this. We're happy to contribute a solution if you could outline something that would be an acceptable approach that we have a chance to get merged.

**Comment by dsherret** [comment]
@joscha I was thinking about this one yesterday! I want to implement it soon. I think probably going to go with what's outlined in https://github.com/dprint/dprint/issues/365#issuecomment-913279108

I think the way to give people the most control over this one is to have the `"includes"` and `"excludes"` globs be used to get a list of files to format and then the `"associations"` in a plugin config can be used to grab out of that list based on a path or glob. It might confuse some users though who wonder why what they specify in `"associations"` isn't being globbed for though, but I will try to make that clear through documentation.

I've basically completed this work in #435. Just need to add tests for this.

**Comment by joscha** [comment]
> I was thinking about this one yesterday! I want to implement it soon. I think probably going to go with what's outlined in [#365 (comment)](https://github.com/dprint/dprint/issues/365#issuecomment-913279108)

that looks pretty good. I think that could work. is it worth giving a priority for each plugin to define the order in which overlapping associations are formatted by each plugin?
The reason I am asking is that for example for our python formatting we format the general code independently from the imports. The import order formatting and splitting is done by a separate tool which we would probably model as a separate plugin, but both would work on the same file extensioons (`*.py`).

cc @groodt @tdeboissiere

**Comment by dsherret** [comment]
> that looks pretty good. I think that could work. is it worth giving a priority for each plugin to define the order in which overlapping associations are formatted by each plugin?

It can be solved by adding a negated glob to the association. For example:

```json
{
  "plugin1": {
    "associations": [
      "./some_folder/**/*.ts",
      "!./some_folder/exclude/**/*.ts",
    ]
  }
  "plugin2": {
    "associations": [
      "./some_folder/exclude/**/*.ts",
    ],
  },
  "includes": [
    "**/*.ts"
  ]
  }
}
```

So that would match every .ts file for `plugin1` except for the ones in `./some_folder/exclude`. Note that the priority is based on how they are ordered in the config file.

> The reason I am asking is that for example for our python formatting we format the general code independently from the imports. The import order formatting and splitting is done by a separate tool which we would probably model as a separate plugin, but both would work on the same file extensioons (*.py).

I see, so you would like to be able to format a single file with two plugins one after the other. That's not possible with this as it always only formats a file with one plugin. I'm wondering if there should be a way to specify a plugin that's a combination of multiple plugins... opened #437, but that's getting kind of complicated. Maybe associations should just allow a file to be captured by multiple plugins and then it formats with both. Opened #438.

**Comment by dsherret** [comment]
I'm wondering though, how hard would it be to combine them into a single plugin that uses both? I wonder how often the scenario of using two plugins to format a file comes up.

I also wonder if there's a way that wasm plugins could be combined by a server like plugins.dprint.dev to create a single wasm plugin from multiple wasm plugins based on the wasm files. It seems like it should be possible.

**Comment by joscha** [comment]
I think the latter yes, multiple, ordered formatting runs. If each plugin
had a priority it would be easy to define a clear order.

**Comment by dsherret** [comment]
This is released now in 0.19.0. Read more here: https://dprint.dev/config/#associations

**Comment by TiceCosmos** [comment]
looks pretty good.
